File: AI/codes/llm-train-python/model/MultiHeadAttention.py
from torch import nn
import torch


class MultiHeadAttention(nn.Module):
    def __init__(self, d_in, d_out, context_length, dropout, num_heads, qkv_bias=False):
        super().__init__()
        assert d_out % num_heads == 0, "d_out must be divisible by num_heads"
        self.d_out = d_out
        self.num_heads = num_heads
        # 将投影维度缩小，以匹配期望的输出维度
        self.head_dim = d_out // num_heads

        self.W_query = nn.Linear(d_in, d_out, bias=qkv_bias)
        self.W_key = nn.Linear(d_in, d_out, bias=qkv_bias)
        self.W_value = nn.Linear(d_in, d_out, bias=qkv_bias)
        # 使用线性层，组合头部输出
        self.out_proj = nn.Linear(d_out, d_out)
        self.dropout = nn.Dropout(dropout)

        self.register_buffer(
            'mask',
            torch.triu(torch.ones(context_length, context_length), diagonal=1)
        )

    def forward(self, x):
        b, num_tokens, d_in = x.shape

        # 张量形状：b, num_tokens, d_out
        keys = self.W_key(x)
        queries = self.W_query(x)
        values = self.W_value(x)

        # 将 d_out 拆成 num_heads * head_dim
        # 张量形状：(b, num_tokens, num_heads, head_dim)
        keys = keys.view(b, num_tokens, self.num_heads, self.head_dim)
        values = values.view(b, num_tokens, self.num_heads, self.head_dim)
        queries = queries.view(b, num_tokens, self.num_heads, self.head_dim)

        # 将张量的形状转置成 (b,num_heads,num_tokens,head_dim)
        keys = keys.transpose(1, 2)
        queries = queries.transpose(1, 2)
        values = values.transpose(1, 2)

        # (b,num_heads,num_tokens,head_dim) * (b,num_heads,head_dim, num_tokens)
        # = (b,num_heads,num_tokens,num_tokens)
        attn_scores = queries @ keys.transpose(2, 3)

        # 掩码被截断到 token 的数量
        mask_bool = self.mask.bool()[:num_tokens, :num_tokens]
        # 填充 -inf，让当前位置不能关注未来 token
        attn_scores.masked_fill_(mask_bool, -torch.inf)

        attn_weights = torch.softmax(attn_scores / keys.shape[-1] ** 0.5, dim=-1)
        attn_weights = self.dropout(attn_weights)

        # (b,num_heads,num_tokens,num_tokens) * (b,num_heads,num_tokens,head_dim)
        # = (b,num_heads,num_tokens,head_dim)
        # ' = (b,num_tokens,num_heads,head_dim)
        context_vec = (attn_weights @ values).transpose(1, 2)

        # 将后最两个维度再合并在一起：num_heads * head_dim = d_out
        # (b,num_tokens,d_out)
        context_vec = context_vec.contiguous().view(b, num_tokens, self.d_out)

        # (b, num_tokens, d_out)
        context_vec = self.out_proj(context_vec)

        return context_vec

File: AI/codes/llm-train-cpp/src/model/MultiHeadAttention.cpp
#include "llm/model/MultiHeadAttention.hpp"

#include "llm/ops.hpp"

#include <stdexcept>

namespace llm {

MultiHeadAttention::MultiHeadAttention(int64_t d_in, int64_t d_out_, int64_t context, int64_t heads,
                                       bool qkv_bias, Device device)
    : d_out(d_out_), num_heads(heads), head_dim(d_out_ / heads), context_length(context),
      W_query(d_in, d_out_, qkv_bias, device), W_key(d_in, d_out_, qkv_bias, device),
      W_value(d_in, d_out_, qkv_bias, device), out_proj(d_out_, d_out_, true, device) {
    if (d_out % num_heads != 0) {
        throw std::runtime_error("d_out must be divisible by num_heads");
    }
}

// 符号：B=batch, T=序列长度, C=d_in(=emb_dim), d_out=输出维度, H=num_heads, Dh=head_dim(=d_out/H)
// x: (B, T, C) -> 返回: (B, T, d_out)
Tensor MultiHeadAttention::forward(const Tensor& x) {
    int64_t B = x.shape()[0], T = x.shape()[1];
    // 线性投影 (B,T,C)->(B,T,d_out)，拆成多头 (B,T,H,Dh)，再转置成 (B,H,T,Dh)
    Tensor q = ops::transpose(ops::reshape(W_query.forward(x), {B, T, num_heads, head_dim}), 1, 2);
    Tensor k = ops::transpose(ops::reshape(W_key.forward(x), {B, T, num_heads, head_dim}), 1, 2);
    Tensor v = ops::transpose(ops::reshape(W_value.forward(x), {B, T, num_heads, head_dim}), 1, 2);
    // k 转置最后两维：(B,H,T,Dh) -> (B,H,Dh,T)
    Tensor kt = ops::transpose(k, 2, 3);
    // 注意力分数 (B,H,T,Dh) @ (B,H,Dh,T) -> (B,H,T,T)
    Tensor scores = ops::batch_matmul(q, kt);
    double scale = 1.0 / std::sqrt(static_cast<double>(head_dim));
    scores = ops::mul_scalar(scores, scale);          // (B,H,T,T) 按 1/sqrt(Dh) 缩放
    scores = ops::causal_mask(scores, T);             // (B,H,T,T) 上三角掩码
    Tensor weights = ops::softmax(scores, -1);        // (B,H,T,T) 末维归一化
    // 加权求和 (B,H,T,T) @ (B,H,T,Dh) -> (B,H,T,Dh)
    Tensor ctx = ops::batch_matmul(weights, v);
    // 合并多头：(B,H,T,Dh) -> (B,T,H,Dh) -> (B,T,d_out)
    ctx = ops::reshape(ops::transpose(ctx, 1, 2), {B, T, d_out});
    return out_proj.forward(ctx);                     // (B,T,d_out) -> (B,T,d_out)
}

std::vector<Tensor*> MultiHeadAttention::parameters() {
    auto p = W_query.parameters();
    auto append = [&](std::vector<Tensor*> q) {
        p.insert(p.end(), q.begin(), q.end());
    };
    append(W_key.parameters());
    append(W_value.parameters());
    append(out_proj.parameters());
    return p;
}

} // namespace llm
